### PR TITLE
Block new strand generation within locks.

### LINKED ISSUE #21232: Panic when a new strand is created inside a lock
**Description:**
Need to panic the program if a new strand is created inside a lock.

for an example :

function lockFunc() {
    lock {
        _ = lockFunc2();
    }
}

functin lockFunc2() {
    worker w1 {
        int n = 10000000;
        int sum = 0;
        foreach var i in 1...n {
            sum += i;
        }
        io:println("sum of first ", n, " positive numbers = ", sum);
    }
}

**Steps to reproduce:**

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**
[ballerina-platform/ballerina-spec#418](https://github.com/ballerina-platform/ballerina-spec/issues/418)

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### LINKED ISSUE #21357: Remove lock from --experimental tag
**Description:**
As we are stabilising lock statement in 2020R1 spec, we need to take the lock statement out from --experimental switch. 

**Steps to reproduce:**

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by dulvinw**
This PR will also disable --experimental tag checking for lock statement since from 1.2 lock statement is stabilized.

Fixes #21357
Fixes #21232

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/worker/WorkerTest.java [hunk 1]
```diff
@@ -347,4 +348,16 @@ public void waitInReturn() {
...
+    @Test(expectedExceptions = BLangRuntimeException.class)
+    public void testFunctionWithWorkerInsideLock() {
+        BValue[] returns = BRunUtil.invoke(result, "testPanicWorkerInsideLock");
+    }
+
+    @Test
+    public void testWorkerInsideLock() {
+        CompileResult result = BCompileUtil.compile("test-src/workers/worker-in-lock.bal");
+        int index = 0;
+        BAssertUtil.validateError(result, index++, 4, 20);
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/worker/WorkerTest.java]: Shall we validate the error message also?

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BLockStore.java [hunk 2]
```diff
@@ -40,9 +40,18 @@ public void addLockToMap(String lockName) {
...
+    public BLock getLockFromMap(String lockName) {
         return globalLockMap.computeIfAbsent(lockName, (k) -> {
             return new BLock();
         });
     }
+
+    public void panicIfInLock(String lockName) {
+        BLock lock = globalLockMap.get(lockName);
+        if (lock != null) {
+            if (!lock.isLockFree()) {
```
Comment (MaryamZi) [inline bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BLockStore.java]: We can combine the checks right?

```suggestion
        if (lock != null && !lock.isLockFree()) {
```

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BallerinaErrors.java [hunk 3]
```diff
@@ -162,6 +163,10 @@ public static ErrorValue createNumericConversionError(Object inputValue, BType i
                 RuntimeErrors.INCOMPATIBLE_SIMPLE_TYPE_CONVERT_OPERATION, inputType, inputValue, targetType));
     }
 
+    public static ErrorValue createAsyncCallInsideLockError() {
```
Comment (MaryamZi) [inline bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BallerinaErrors.java]: This method is only called in BLockStore.java right? Shall we move the method there?

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BallerinaErrors.java [hunk 4]
```diff
@@ -61,6 +61,7 @@
     public static final String GENERATE_PKG_START = "___start_";
     public static final String GENERATE_PKG_STOP = "___stop_";
     public static final String GENERATE_OBJECT_CLASS_PREFIX = ".$value$";
+    public static final String ASYNC_CALL_INSIDE_LOCK = "AsyncCallInsideLock";
```
Comment (MaryamZi) [inline bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BallerinaErrors.java]: Shall we move this constant to BallerinaErrorReasons.java and prefix it with "{ballerina}"? (Check BallerinaErrorReasons for examples).

And I guess we should suffix it with "Error"?

```suggestion
    public static final String ASYNC_CALL_INSIDE_LOCK = "AsyncCallInsideLockError";
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java [hunk 5]
```diff
@@ -1998,7 +2000,12 @@ public void visit(BLangInvocation invocationExpr) {
         }
 
         if (invocationExpr.actionInvocation || invocationExpr.async) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: Is it correct to do this for all action invocations too? They may or may not be async right?

File: compiler/ballerina-lang/src/main/resources/compiler.properties [hunk 6]
```diff
@@ -20,6 +20,9 @@
 # Compiler warning messages
 # -------------------------
 
+error.usage.of.async.call.within.lock.is.prohibited=\
+  usage of async call ''{0}'' inside of lock statment is prohibited
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/resources/compiler.properties]: Can we differentiate between this error being cause due to start vs named worker?

Shall we also reword this to something like
```suggestion
  cannot use an async call inside a lock statement
```
and 
```suggestion
  cannot use a named worker inside a lock statement
```

@hasithaa, @pubudu91, thoughts?

There's a typo in "statment" btw.

File: tests/jballerina-unit-test/src/test/resources/test-src/workers/worker-in-lock.bal [hunk 7]
```diff
@@ -0,0 +1,9 @@
+function testWorkerInsideLock() {
+    lock {
+        fork {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/workers/worker-in-lock.bal]: Can we add a test for start too?

Comment (dulvinw) [reply tests/jballerina-unit-test/src/test/resources/test-src/workers/worker-in-lock.bal]: Will add

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/worker/WorkerTest.java (hunk 1 above)
Comment (dulvinw) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/worker/WorkerTest.java]: +1 will do

File: compiler/ballerina-lang/src/main/resources/compiler.properties (hunk 6 above)
Comment (dulvinw) [reply compiler/ballerina-lang/src/main/resources/compiler.properties]: In case of start the same thing should happen as well right?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java (hunk 5 above)
Comment (dulvinw) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: I'll recheck and see.

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BLock.java [hunk 8]
```diff
@@ -61,11 +61,11 @@ public synchronized void unlock() {
...
     }
 
-    private boolean isLockFree() {
+    public boolean isLockFree() {
         return this.current.isEmpty();
     }
 
-    private boolean lockedBySameContext(Strand ctx) {
+    public boolean lockedBySameContext(Strand ctx) {
         return this.current.getLast() == ctx;
```
Comment (hasithaa) [inline bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BLock.java]: Shouldn't this be a recursive check?

File: compiler/ballerina-lang/src/main/resources/compiler.properties (hunk 6 above)
Comment (hasithaa) [reply compiler/ballerina-lang/src/main/resources/compiler.properties]: +1 to reword. I think having two error messages is good.

File: tests/jballerina-unit-test/src/test/resources/test-src/workers/worker-in-lock.bal (hunk 7 above)
Comment (hasithaa) [reply tests/jballerina-unit-test/src/test/resources/test-src/workers/worker-in-lock.bal]: Also Add test case for nested fork statements.

File: tests/jballerina-unit-test/src/test/resources/test-src/workers/workers.bal [hunk 9]
```diff
@@ -542,3 +542,15 @@ function waitInReturn() returns any {
 
     return wait {w1, w2};
 }
+
+function testPanicWorkerInsideLock() {
+    lock {
+        panicWorkerInsideLock();
```
Comment (hasithaa) [inline tests/jballerina-unit-test/src/test/resources/test-src/workers/workers.bal]: Make this work for multiple levels.

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BLock.java (hunk 8 above)
Comment (dulvinw) [reply bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BLock.java]: Since we only create a new strand when we hit a worker or start, enough to check the current parent.

File: tests/jballerina-unit-test/src/test/resources/test-src/workers/worker-in-lock.bal (hunk 7 above)
Comment (dulvinw) [reply tests/jballerina-unit-test/src/test/resources/test-src/workers/worker-in-lock.bal]: +1

File: compiler/ballerina-lang/src/main/resources/compiler.properties (hunk 6 above)
Comment (dulvinw) [reply compiler/ballerina-lang/src/main/resources/compiler.properties]: will fix

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java [hunk 10]
```diff
@@ -1232,9 +1233,10 @@ public void visit(BLangWhile whileNode) {
     @Override
     public void visit(BLangLock lockNode) {
 
-        checkExperimentalFeatureValidity(ExperimentalFeatures.LOCK, lockNode.pos);
         this.checkStatementExecutionValidity(lockNode);
+        this.withinLockBlock = true;
         lockNode.body.stmts.forEach(e -> analyzeNode(e, env));
+        this.withinLockBlock = false;
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: A lock-stmt can be written inside another lock-stmt right? In that case, I think we should maintain the previous `this.withinLockBlock` before setting `this.withinLockBlock` to true at L1237 and reset the previous value here.

Or something like the following may not fail?
```ballerina
function testFunc() {
    lock {
        lock {
            
        }

        fork {
            worker w1 {
                
            }
        }
    }
}
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java [hunk 11]
```diff
@@ -1998,8 +2000,24 @@ public void visit(BLangInvocation invocationExpr) {
...
-            validateActionInvocation(invocationExpr.pos, invocationExpr);
+            if (invocationExpr.async && this.withinLockBlock) {
+                logAsyncCallWithinLockError(invocationExpr);
+            } else {
+                validateActionInvocation(invocationExpr.pos, invocationExpr);
+            }
+        }
+    }
+
+    private void logAsyncCallWithinLockError(BLangInvocation invocationExpr) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: Do we need a separate method?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java [hunk 12]
```diff
@@ -1998,8 +2000,24 @@ public void visit(BLangInvocation invocationExpr) {
         }
 
         if (invocationExpr.actionInvocation || invocationExpr.async) {
-            validateActionInvocation(invocationExpr.pos, invocationExpr);
+            if (invocationExpr.async && this.withinLockBlock) {
+                logAsyncCallWithinLockError(invocationExpr);
+            } else {
+                validateActionInvocation(invocationExpr.pos, invocationExpr);
+            }
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: Shall we switch the checks and return early?
```suggestion
            if (!invocationExpr.async || !this.withinLockBlock) {
                validateActionInvocation(invocationExpr.pos, invocationExpr);
                return;
            }
            // log the error here
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java [hunk 13]
```diff
@@ -1998,8 +2000,24 @@ public void visit(BLangInvocation invocationExpr) {
...
+    private void logAsyncCallWithinLockError(BLangInvocation invocationExpr) {
+        DiagnosticCode code;
+
+        if (invocationExpr.functionPointerInvocation) {
+            code = DiagnosticCode.USAGE_OF_WORKER_WITHIN_LOCK_IS_PROHIBITED;
+        } else {
+            code = DiagnosticCode.USAGE_OF_START_WITHIN_LOCK_IS_PROHIBITED;
         }
+
+        dlog.error(invocationExpr.pos, code);
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: We can simplify this using a ternary operator?
```suggestion
        dlog.error(invocationExpr.pos, invocationExpr.functionPointerInvocation ?  
                   DiagnosticCode.USAGE_OF_WORKER_WITHIN_LOCK_IS_PROHIBITED : 
                   DiagnosticCode.USAGE_OF_START_WITHIN_LOCK_IS_PROHIBITED);
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java [hunk 14]
```diff
@@ -2000,7 +2003,14 @@ public void visit(BLangInvocation invocationExpr) {
         }
 
         if (invocationExpr.actionInvocation || invocationExpr.async) {
-            validateActionInvocation(invocationExpr.pos, invocationExpr);
+            if (!invocationExpr.async || !this.withinLockBlock) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]:
```suggestion
            if (invocationExpr.actionInvocation || !this.withinLockBlock) {
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java [hunk 15]
```diff
@@ -2000,7 +2003,14 @@ public void visit(BLangInvocation invocationExpr) {
...
+                validateActionInvocation(invocationExpr.pos, invocationExpr);
+                return;
+            }
+
+            dlog.error(invocationExpr.pos, invocationExpr.functionPointerInvocation ? 
+            DiagnosticCode.USAGE_OF_WORKER_WITHIN_LOCK_IS_PROHIBITED : 
+            DiagnosticCode.USAGE_OF_START_WITHIN_LOCK_IS_PROHIBITED);
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: Shall we indent these two lines to align with the beginning of invocationExpr.pos?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java (hunk 14 above)
Comment (dulvinw) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: Fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java (hunk 13 above)
Comment (dulvinw) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: Fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java (hunk 12 above)
Comment (dulvinw) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: Fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java (hunk 10 above)
Comment (dulvinw) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: Fixed

File: tests/jballerina-unit-test/src/test/resources/test-src/workers/workers.bal (hunk 9 above)
Comment (dulvinw) [reply tests/jballerina-unit-test/src/test/resources/test-src/workers/workers.bal]: FIxed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java (hunk 5 above)
Comment (dulvinw) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: Fixed

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BallerinaErrors.java (hunk 3 above)
Comment (dulvinw) [reply bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BallerinaErrors.java]: Fixed
